common/src/main/java/io/pravega/common/util/SortedDeque.java

```diff
@@ -0,0 +1,269 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.base.Preconditions;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Deque-like data structure that accepts elements in sorted order and provides optimized operations based on that.
+ * Note: This class was inspired by java.util.ArrayDeque, however that class does not allow subclassing outside of its package.
+ */
+@NotThreadSafe
+@SuppressWarnings("unchecked")
+public class SortedDeque<E extends SortedIndex.IndexEntry> {
+    //region Members
+
+    private static final int MIN_INITIAL_CAPACITY = 8;
+    private transient SortedIndex.IndexEntry[] elements;
+    private transient int lastSlotIndex;
+    private transient int head;
+    private transient int tail;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the SortedDeque class.
+     */
+    public SortedDeque() {
+        this(MIN_INITIAL_CAPACITY);
+    }
+
+    /**
+     * Creates a new instance of the SortedDeque class.
+     *
+     * @param initialCapacity The initial capacity.
+     */
+    public SortedDeque(int initialCapacity) {
+        this.elements = new SortedIndex.IndexEntry[Math.max(MIN_INITIAL_CAPACITY, initialCapacity)];
+        this.lastSlotIndex = this.elements.length - 1;
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Adds an item to the end of the Deque.
+     *
+     * @param item The Item to add
+     * @throws IllegalArgumentException If the item's key is not greater than the key of the last item in the Deque.
+     */
+    public void addLast(E item) {
+        SortedIndex.IndexEntry last = this.elements[(this.tail - 1) & this.lastSlotIndex];
```

Comment:
It's alright. In that case it will not be needed since `head == tail` and all ops will fail. All these bitwise operations are there because the modulo operator is too slow. Check out the base implementation in `java.util.ArrayDeque`.